protocols/request-response/src/lib.rs

```diff
@@ -443,7 +458,7 @@ where
 
         if let Some(request) = self.try_send_request(peer, request) {
             self.pending_events.push_back(ToSwarm::Dial {
-                opts: DialOpts::peer_id(*peer).build(),
+                opts: DialOpts::peer_id(*peer).addresses(addresses).build(),
```

Comment:
By using `DialOpts::addresses` wouldnt this cause an issue where if the address is blank that it would not extend through the behaviour for addresses? If so, we could probably check to see if its empty and use `DialOpts::peer_id(*peer).build()` if it is, otherwise use `DialOpts::peer_id(*peer).addresses(addresses).build()`. Thoughts?

Reply:
Completely forgot about that annoying pitfall, added `.extend_addresses_through_behaviour()` so it populates addresses from both places, thanks for catching this!

Reply:
I was considering suggesting `extend_addresses_through_behaviour` but did not know if if you wanted to only explicitly dial those addresses and not ones supplied by other behaviours and if that was the case, checking the vector might be better.

Reply:
I do not have a hard requirement for that, but don't have strong opinion either way. Using `extend_addresses_through_behaviour` seemed more useful to me.

Reply:
I think it's slightly more intuitive if in case of `_with_addresses` only the provided addresses are used. 

How about we have a private function `send_request_with_opts` that is used with the two different `DialOpts` variants?
I think that would be a bit cleaner than checking the vector length.

Reply:
What is the problem we're trying to solve by this? I think it might be useful to have both addresses at the same time. Depending on use case it also prevents caller from checking the number of addresses if they may have an empty list in their context.

https://github.com/libp2p/rust-libp2p/pull/5692 was meant to provide full control at the cost of increased complexity and breaking API. This version provides a simple and misuse-resistant additional method instead.

Reply:
My idea was to have a clear separation between using `send_request`, which relies on an address being present in the address-store/ other behaviors, and using `send_request_with_addresses` to send a request over a specific set of addresses (allowing e.g. to select what transport should be used). 

However, thinking about it now again, we might want to make the address-store opt-out in the future anyway, in which case there might be more use-cases for `send_request_with_addresses` that also want to leverage other behavior's address info.

So currently version is fine by me.